torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -8010,10 +8008,26 @@ def buffer_comm_hook(ddp, named_buffers):
                         for tensor in model_no_hook_buffers:
                             dist.all_reduce(tensor)
 
-                    self._verify_buffers_equal(model_ddp, model_ddp_no_hook)
+                    # if return_futures, they are only awaited on by DDP
+                    # at the end of the backwards pass for maximum overlap.
+                    if not return_futures:
+                        self._verify_buffers_equal(model_ddp, model_ddp_no_hook)
                     loss_hook.backward()
                     loss_no_hook.backward()
+                    # Note that when custom hooks return futures, this comparison is not expected to work when hook run location is pre-forward pass. This is because the hook does async communication
+                    # and forward pass modifies the buffer without appropriate synchronization. Therefore, if returning futures from custom buffer hooks, it is advised to set hook run location to post
+                    # forward.
+                    if return_futures and hook_run_location == hook_post_fwd:
```

Comment:
hook_run_location does not have to be 'hook_post_fwd'?

Reply:
It actually does, if you see the comments on L8017. Basically if we run the hook _before_ forward and return futures, then the user must perform appropriate synchronization in their forward pass for buffers to be equal. This is because the forward pass would use buffers without appropriate synchronization with the NCCL stream if we are returning futures. 

For the test we can manually change it to synchronize, but I think the best bet here is to document this and let users know they must explicitly synchronize if they want to use `return_futures` and `hook_run_location == hook_pre_fwd` (this is already documented on 1289).